includes/data-events/class-data-events.php

```diff
@@ -0,0 +1,230 @@
+	 *
+	 * @return void|WP_Error Error if action already registered.
+	 */
+	public static function register_action( $action_name ) {
+		if ( isset( self::$actions[ $action_name ] ) ) {
+			return new WP_Error( 'action_already_registered', __( 'Action already registered.', 'newspack' ) );
+		}
+		self::$actions[ $action_name ] = [];
+	}
+
+	/**
+	 * Register a handler for a triggerable action.
+	 *
+	 * @param callable $handler     Action handler.
+	 * @param string   $action_name Action name.
+	 *
+	 * @return void|WP_Error Error if action not registered, handler already registered or is not callable.
+	 */
+	public static function register_handler( $handler, $action_name = null ) {
+		/** If first argument is callable, treat as global handler. */
+		if ( empty( $action_name ) ) {
+			self::$global_handlers[] = $handler;
+			return;
+		}
+		if ( ! isset( self::$actions[ $action_name ] ) ) {
+			return new WP_Error( 'action_not_registered', __( 'Action not registered.', 'newspack' ) );
+		}
+		if ( ! is_callable( $handler ) ) {
+			return new WP_Error( 'handler_not_callable', __( 'Handler is not callable.', 'newspack' ) );
+		}
+		if ( in_array( $handler, self::$actions[ $action_name ], true ) ) {
+			return new WP_Error( 'handler_already_registered', __( 'Handler already registered.', 'newspack' ) );
+		}
+		self::$actions[ $action_name ][] = $handler;
+	}
+
+	/**
+	 * Get a list of all registered actions.
+	 *
+	 * @return string[] Registered actions.
+	 */
+	public static function get_actions() {
+		return array_keys( self::$actions );
+	}
+
+	/**
+	 * Get a list of all registered action handlers.
+	 *
+	 * @param string $action_name Action name.
+	 *
+	 * @return callable[] Registered action handlers.
+	 */
+	public static function get_action_handlers( $action_name ) {
+		if ( ! isset( self::$actions[ $action_name ] ) ) {
+			return [];
+		}
+		return self::$actions[ $action_name ];
+	}
+
+	/**
+	 * Whether an action is registered.
+	 *
+	 * @param string $action_name The action name.
+	 *
+	 * @return bool
+	 */
+	public static function is_action_registered( $action_name ) {
+		return isset( self::$actions[ $action_name ] );
+	}
+
+	/**
+	 * Dispatch an action event.
+	 *
+	 * @param string  $action_name   Action name.
+	 * @param array   $data          Data to pass to the action.
+	 * @param boolean $use_client_id Whether to use the session's client ID. Default is true.
+	 *
+	 * @return void|WP_Error Error if action not registered.
+	 */
+	public static function dispatch( $action_name, $data, $use_client_id = true ) {
+		if ( ! self::is_action_registered( $action_name ) ) {
+			return new WP_Error( 'newspack_data_events_action_not_registered', __( 'Action not registered.', 'newspack' ) );
+		}
+
+		$timestamp = time();
+		$client_id = null;
+		if ( $use_client_id ) {
+			$client_id = Reader_Activation::get_client_id();
+		}
+
+		do_action( 'newspack_data_event_dispatch', $action_name, $timestamp, $data, $client_id );
```

Comment:
Do we have any immediate use for these hooks? If not, I'd say drop them... this is precisely the function we want to execute fast and adding these hooks here gives room for developers to defeat this purpose.

If we do have a use for these hooks, document them and add this warning that any heavy load should be done in the action handler and not here

Reply:
This is meant for the extensibility of the tool, not as an actual third-party action handler. There are 2 use cases off the top of my head, but in terms of "extensibility" _anything_ is possible 😅 

1. The webhooks functionality will have its own async handling, because of its particular error-handling flow. It will very likely hook into the dispatch so it creates webhook requests accordingly.
2. If we ever decide to print front-end js for pushing analytics events, this would have to be in the dispatch execution. Example of a very hypothetical code:

```php
add_action(
	'newspack_data_event_dispatch',
	function( $action_name, $timestamp, $data, $client_id ) {
		add_action(
			'wp_footer',
			function() {
				echo '<script>dataLayer.push({"event": "' . esc_attr( $action_name ) . '"});</script>';
			}
		)
	}
	10,
	4
);
```

Reply:
For number 1, is there any reason why webhooks couldn't do their thing in the handler? It would trigger its async call from an async call...

Reply:
This is something yet to experiment with actual code, but since a request is going to be a `post`, it's more suitable to use its scheduling functionality for this. Each registered webhook endpoint will create a request, which will be rescheduled to send the data until it returns `2xx`.

It feels unnecessary to hook the creation of the request as an async handler if the handling is not happening at this point.